### PR TITLE
feat(gpu): add gpu array type in hl api

### PR DESCRIPTION
**Pull request by agnesLeroy**
closes: _please link all relevant issues_

### Check-list:

* [ ] Tests for the changes have been added (for bug fixes / features)
* [ ] Docs have been added / updated (for bug fixes / features)
* [ ] Relevant issues are marked as resolved/closed, related issues are linked in the description
* [ ] Check for breaking changes (including serialization changes) and add them to commit message following the conventional commit [specification][conventional-breaking]

[conventional-breaking]: https://www.conventionalcommits.org/en/v1.0.0/#commit-message-with-description-and-breaking-change-footer

### CODE REVIEW

File: tfhe/src/integer/gpu/ciphertext/boolean_value.rs [hunk 1]
```diff
@@ -188,3 +190,9 @@ impl AsMut<CudaUnsignedRadixCiphertext> for CudaBooleanBlock {
         &mut self.0
     }
 }
+impl Clone for CudaBooleanBlock {
+    fn clone(&self) -> Self {
+        let streams = CudaStreams::new_multi_gpu();
+        Self(self.0.duplicate(&streams))
+    }
+}
```
Comment (tmontaigu) [inline tfhe/src/integer/gpu/ciphertext/boolean_value.rs:198]: No sure we should impl that, here as CudaBooleanBlock in as integer thing and, CudaUnsined/SignedRadix do not impl Clone

Maybe we have to the same thing as the integer backend  (have a CudaOwned(Vec<..>) and impl Clone for it

Or move the `impl Clone for CudaBooleanBlock` to integer (and  probably impl it also for Radix types)

File: tfhe/src/high_level_api/array/gpu/integers.rs [hunk 2]
```diff
@@ -0,0 +1,648 @@
+//! This module contains the implementations of the FheUint array and FheInt array backend
+//! where the values and computations are always done on CPU
```
Comment (tmontaigu) [inline tfhe/src/high_level_api/array/gpu/integers.rs:2]:
```suggestion
where the values and computations are always done on CPU
```

File: tfhe/src/integer/gpu/ciphertext/boolean_value.rs (hunk 1 above)
Comment (agnesLeroy) [reply tfhe/src/integer/gpu/ciphertext/boolean_value.rs:198]: Ok I'll follow the same logic with GpuOwned, etc for Booleans.

File: tfhe/src/high_level_api/array/gpu/integers.rs (hunk 2 above)
Comment (agnesLeroy) [reply tfhe/src/high_level_api/array/gpu/integers.rs]: If I remove `//!` the line won't be a comment anymore, but I have to replace CPU with GPU, is it what you had in mind?

File: tfhe/src/integer/gpu/ciphertext/boolean_value.rs (hunk 1 above)
Comment (agnesLeroy) [reply tfhe/src/integer/gpu/ciphertext/boolean_value.rs]: I introduced GpuBooleanSlice, GpuBooleanSliceMut, GpuBooleanOwned to do it, the change is pushed.

File: tfhe/src/high_level_api/array/gpu/integers.rs (hunk 2 above)
Comment (tmontaigu) [reply tfhe/src/high_level_api/array/gpu/integers.rs]: haha no my suggestion is wrong, I meant to change CPU to GPU 😓

**Review by tmontaigu — APPROVED** [review]

**Pull request merged**